multinode/adapter.go

```diff
@@ -0,0 +1,302 @@
+	blockNumber := head.BlockNumber()
+	totalDifficulty := head.GetTotalDifficulty()
+	if !CtxIsHeathCheckRequest(ctx) {
+		m.highestUserObservations.BlockNumber = max(m.highestUserObservations.BlockNumber, blockNumber)
+		m.highestUserObservations.TotalDifficulty = MaxTotalDifficulty(m.highestUserObservations.TotalDifficulty, totalDifficulty)
+	}
+	select {
+	case <-requestCh: // no need to update latestChainInfo, as rpcMultiNodeAdapter already started new life cycle
+		return
+	default:
+		m.latestChainInfo.BlockNumber = blockNumber
+		m.latestChainInfo.TotalDifficulty = totalDifficulty
+	}
+}
+
+func (m *Adapter[RPC, HEAD]) OnNewFinalizedHead(ctx context.Context, requestCh <-chan struct{}, head HEAD) {
+	if !head.IsValid() {
+		return
+	}
+
+	m.chainInfoLock.Lock()
+	defer m.chainInfoLock.Unlock()
+	if !CtxIsHeathCheckRequest(ctx) {
+		m.highestUserObservations.FinalizedBlockNumber = max(m.highestUserObservations.FinalizedBlockNumber, head.BlockNumber())
+	}
+	select {
+	case <-requestCh: // no need to update latestChainInfo, as rpcMultiNodeAdapter already started new life cycle
+		return
+	default:
+		m.latestChainInfo.FinalizedBlockNumber = head.BlockNumber()
+	}
+}
+
+// MakeQueryCtx returns a context that cancels if:
+// 1. Passed in ctx cancels
+// 2. Passed in channel is closed
+// 3. Default timeout is reached (queryTimeout)
+func MakeQueryCtx(ctx context.Context, ch services.StopChan, timeout time.Duration) (context.Context, context.CancelFunc) {
+	var chCancel, timeoutCancel context.CancelFunc
+	ctx, chCancel = ch.Ctx(ctx)
+	ctx, timeoutCancel = context.WithTimeout(ctx, timeout)
+	cancel := func() {
+		chCancel()
+		timeoutCancel()
+	}
+	return ctx, cancel
+}
+
+func (m *Adapter[RPC, HEAD]) AcquireQueryCtx(parentCtx context.Context, timeout time.Duration) (ctx context.Context, cancel context.CancelFunc,
+	chStopInFlight chan struct{}, raw *RPC) {
+	// Need to wrap in mutex because state transition can cancel and replace context
+	m.stateMu.RLock()
+	chStopInFlight = m.chStopInFlight
+	cp := *m.rpc
+	raw = &cp
+	m.stateMu.RUnlock()
+	ctx, cancel = MakeQueryCtx(parentCtx, chStopInFlight, timeout)
+	return
+}
+
+func (m *Adapter[RPC, HEAD]) UnsubscribeAllExcept(subs ...Subscription) {
+	m.subsSliceMu.Lock()
+	keepSubs := map[Subscription]struct{}{}
+	for _, sub := range subs {
+		keepSubs[sub] = struct{}{}
+	}
+
+	var unsubs []Subscription
+	for sub := range m.subs {
+		if _, keep := keepSubs[sub]; !keep {
+			unsubs = append(unsubs, sub)
+		}
+	}
+	m.subsSliceMu.Unlock()
+
+	for _, sub := range unsubs {
+		sub.Unsubscribe()
+	}
+}
+
+// CancelInflightRequests closes and replaces the chStopInFlight
+func (m *Adapter[RPC, HEAD]) CancelInflightRequests() {
+	m.stateMu.Lock()
+	defer m.stateMu.Unlock()
+	close(m.chStopInFlight)
+	m.chStopInFlight = make(chan struct{})
+}
+
+// GetChStopInflight provides a convenience helper that mutex wraps a
+// read to the chStopInFlight
+func (m *Adapter[RPC, HEAD]) GetChStopInflight() chan struct{} {
```

Comment:
Do we need all GetChStopInflight, MakeQueryCtx and AcquireQueryCtx methods?

Reply:
We still need them all for now. GetChStopInflight is used in EVM since it implements its own acquireQueryCtx due to containing both http and ws endpoints to be copied. For new chain integrations, the default implemented by the adapter will be good enough until we add ws support throughout.

Reply:
Not anymore! Got rid of GetChStopInflight and unexported makeQueryCtx since it's only used in AcquireQueryCtx.